File: android/app/src/main/java/io/twentysixty/mobileagent/FileChunkGeneratorModule.java
package io.twentysixty.mobileagent;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableNativeArray;
import com.facebook.react.bridge.WritableArray;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import android.util.Log;

public class FileChunkGeneratorModule extends ReactContextBaseJavaModule {
    FileChunkGeneratorModule(ReactApplicationContext context) {
        super(context);
    }

    @Override
    public String getName() {
        return "FileChunkGeneratorModule";
    }

    @ReactMethod
    public void createChunks(String filePath, String outputFilePathPrefix, int chunkSize, final Promise promise) {
        Log.d("FileChunkGeneratorModule", "createChunks of size " + chunkSize + " for file: " + filePath
                + " and output directory: " + outputFilePathPrefix);

        int chunkCount = 0;
        List<String> outputPaths = new ArrayList<String>();

        try {
            File inputFile = new File(filePath);
            FileInputStream fis = new FileInputStream(inputFile);

            while(fis.available() > 0) {
                byte [] container = new byte[fis.available() < chunkSize ? fis.available() : chunkSize];
                int currentChunkSize = fis.read(container);
                Log.d("FileChunkGeneratorModule", "creating chunk " + Integer.toString(chunkCount)
                        + " with size: " + currentChunkSize);
                String chunkFilePath = outputFilePathPrefix + '.' + Integer.toString(chunkCount++);
                FileOutputStream fos = new FileOutputStream(chunkFilePath);
                fos.write(container);
                fos.close();
                outputPaths.add(chunkFilePath);
                Log.d("FileChunkGeneratorModule", "created chunk " + Integer.toString(chunkCount));
            }
        } catch(Exception e) {
            Log.d("FileChunkGeneratorModule", "Error: " + e.getMessage());
        }

        String[] returnArray = outputPaths.toArray(new String[0]);

        WritableArray promiseArray = Arguments.createArray();
        for (int i = 0 ; i < returnArray.length; i++) {
            promiseArray.pushString(returnArray[i]);
        }

        promise.resolve(promiseArray);

    }

    @ReactMethod
    public void readChunk(String filePath, double offset, double length, final Promise promise) {
        Log.d("FileChunkGeneratorModule", "read chunk for file: " + filePath
                + " at offset " + offset + "with length " + length);
        try {
            File inputFile = new File(filePath);
            FileInputStream fis = new FileInputStream(inputFile);

            byte [] data = new byte[(int) length];
            fis.skip((long) offset);
            Log.d("FileChunkGeneratorModule", "reading chunk");
            fis.read(data);
            Log.d("FileChunkGeneratorModule", "chunk succesfully read");
            WritableArray result = Arguments.createArray();
            for (byte b : data) {
                result.pushInt((int) b);
            }
            promise.resolve(result);
        } catch(Exception e) {
            Log.d("FileChunkGeneratorModule", "Error: " + e.getMessage());
            promise.reject("-1", e.getMessage());
        }
    }

}

File: modules/react-native-local-native-modules/android/src/main/java/com/reactnativelocalnativemodules/FileChunkGeneratorModule.kt
package com.reactnativelocalnativemodules

import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.module.annotations.ReactModule
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream

@ReactModule(name = FileChunkGeneratorModule.NAME)
class FileChunkGeneratorModule(reactContext: ReactApplicationContext) :
  NativeFileChunkGeneratorSpec(reactContext) {

    override fun createChunks(
        filePath: String,
        outputFilePathPrefix: String,
        chunkSize: Double,
        promise: Promise
    ) {
        Log.d(
            "FileChunkGeneratorModule",
            ("createChunks of size " + chunkSize.toInt() + " for file: " + filePath
                    + " and output directory: " + outputFilePathPrefix)
        )

        var chunkCount = 0
        val outputPaths: MutableList<String> = ArrayList()

        try {
            val inputFile = File(filePath)
            val fis = FileInputStream(inputFile)

            while (fis.available() > 0) {
                val container =
                    ByteArray(if (fis.available() < chunkSize.toInt()) fis.available() else chunkSize.toInt())
                val currentChunkSize = fis.read(container)
                Log.d(
                    "FileChunkGeneratorModule",
                    ("creating chunk $chunkCount with size: $currentChunkSize")
                )
                val chunkFilePath = outputFilePathPrefix + '.' + chunkCount++.toString()
                val fos = FileOutputStream(chunkFilePath)
                fos.write(container)
                fos.close()
                outputPaths.add(chunkFilePath)
                Log.d("FileChunkGeneratorModule", "created chunk $chunkCount")
            }
        } catch (e: Exception) {
            Log.d("FileChunkGeneratorModule", "Error: " + e.message)
        }

        val returnArray = outputPaths.toTypedArray<String>()

        val promiseArray = Arguments.createArray()
        for (i in returnArray.indices) {
            promiseArray.pushString(returnArray[i])
        }

        promise.resolve(promiseArray)
    }

    override fun readChunk(filePath: String, offset: Double, length: Double, promise: Promise) {
        Log.d(
            "FileChunkGeneratorModule", ("read chunk for file: " + filePath
                    + " at offset " + offset + "with length " + length)
        )
        try {
            val inputFile = File(filePath)
            val fis = FileInputStream(inputFile)

            val data = ByteArray(length.toInt())
            fis.skip(offset.toLong())
            Log.d("FileChunkGeneratorModule", "reading chunk")
            fis.read(data)
            Log.d("FileChunkGeneratorModule", "chunk succesfully read")
            val result = Arguments.createArray()
            for (b in data) {
                result.pushInt(b.toInt())
            }
            promise.resolve(result)
        } catch (e: Exception) {
            Log.d("FileChunkGeneratorModule", "Error: " + e.message)
            promise.reject("-1", e.message)
        }
    }

    companion object {
    const val NAME = "FileChunkGenerator"
  }
}
